datalad/interface/unlock.py

```diff
@@ -69,97 +72,76 @@ def __call__(
-                    ap['message'] = "not annex'ed, nothing to unlock"
-                    ap.update(res_kwargs)
-                    yield ap
-                continue
-
-            # only files in annex with their content present:
-            files = [ap['path'] for ap in content]
-            to_unlock = []
-            for ap, under_annex, has_content in \
-                zip(content,
-                    ds.repo.is_under_annex(files),
-                    ds.repo.file_has_content(files)):
-
-                # TODO: what about directories? Make sure, there is no
-                # situation like no file beneath with content or everything in
-                # git, that leads to a CommandError
-                # For now pass to annex:
-                from os.path import isdir
-                if isdir(ap['path']):
-                    to_unlock.append(ap)
-                    continue
-
-                # Note, that `file_has_content` is (planned to report) True on
-                # files in git. Therefore order matters: First check for annex!
-                if under_annex:
-                    if has_content:
-                        to_unlock.append(ap)
-                    # no content, no unlock:
-                    else:
-                        ap['status'] = 'impossible'
-                        ap['message'] = "no content present, can't unlock"
-                        ap.update(res_kwargs)
-                        yield ap
-                # file in git, no unlock:
+            has_content = res.get("has_content")
+            if has_content:
+                to_unlock[res["parentds"]].append(res["path"])
+            elif paths_nondir and Path(res["path"]) in paths_nondir:
+                if has_content is False:
+                    msg = "no content present"
+                elif res["state"] == "untracked":
+                    msg = "untracked"
                 else:
-                    ap['status'] = 'notneeded'
-                    ap['message'] = "not controlled by annex, nothing to unlock"
-                    ap.update(res_kwargs)
-                    yield ap
-
-            # don't call annex-unlock with no path, if this is this case because
-            # nothing survived the filtering above
-            if content and not to_unlock:
-                continue
+                    # TODO: Anyway to distinguish between a regular git file
+                    # and an unlocked annex file?
+                    msg = "non-annex file"
+                yield get_status_dict(
+                    status="impossible",
+                    path=res["path"],
+                    type="file",
+                    message="{}; cannot unlock".format(msg),
+                    **res_kwargs)
 
-            for r in ds.repo.unlock([ap['path'] for ap in to_unlock]):
+        # Do the actual unlocking.
+        for ds_path, files in to_unlock.items():
+            ds = Dataset(ds_path)
+            for r in ds.repo.unlock(files):
```

Comment:
We have to recode file names to be relative to the repository. Otherwise, repos in symlinked paths will not work.

`f.relative_to(ds.pathobj) for f in to_unlock...`

if `f`s are `Path` instances.

If anyhow possible, it would be great to also incorporate `AnnexRepo.unlock()` into this RF:

- [ ] shed `normalize_paths` deocrator -- which prevents actual generator-like behavior
- [ ] call `git annex unlock` with `--json --json-error-messages` and relay the full results, rather than manual parsing
- [ ] strip `options` argument -- there are no non-generic options supported by `git annex unlock`, our API should reflect that, and not make `unlock`-specific exceptions.
- [ ] strip `AnnexRepo.lock()` from the codebase -- unused, untested, no high-level API companion (a simple `save` does the job)

Reply:
> We have to recode file names to be relative to the repository. Otherwise, repos in symlinked paths will not work.

Quick testing with a repo that had a symlink in the leading path, things seem to work, but I'm may not be testing the right case.  I know you've dealt with this issue a good amount and I have no reason not to use a relative path, so I'll go with your suggestion.

> If anyhow possible, it would be great to also incorporate `AnnexRepo.unlock()` into this RF [...]

Hmm looking into that hadn't occurred to me.  Will do.

Reply:
Started to look into AnnexRepo.unlock()...

> * [ ]  shed `normalize_paths` deocrator -- which prevents actual generator-like behavior

I'm for dropping `normalize_paths` (here and in general), but isn't any useful generator-like behavior also prevented by `git annex unlock` not supporting `--batch`?  So, assuming we use _run_annex_command_json(), we get a list of json objects, and it doesn't seem particularly useful to, say, yield those in turn rather than just return the list.

> * [ ]  call `git annex unlock` with `--json --json-error-messages` and relay the full results, rather than manual parsing

I agree we should use --json.  --json-error-messages _sounds_ like it'd be useful, but I'm not sure it is in its current form:

```sh
#!/bin/sh

set -eu

cd $(mktemp -dt gx-XXXXXXX)

git init
git annex init
git annex unlock --json --json-error-messages notthere
```

```
Initialized empty Git repository in /tmp/gx-rHLSVpQ/.git/
init  ok
(recording state in git...)
git-annex: notthere not found
git-annex: unlock: 1 failed
```

> * [ ]  strip `options` argument -- there are no non-generic options supported by `git annex unlock`, our API should reflect that, and not make `unlock`-specific exceptions.

Sounds fine to me.

> * [ ]  strip `AnnexRepo.lock()` from the codebase -- unused, untested, no high-level API companion (a simple `save` does the job)

Also OK with me.

Reply:
OK, I've pushed updates to AnnexRepo.lock().  It only partially addresses your suggestions.

It

  * drops lock()
  * removes unlock's `options` parameter.
  * uses `--json` (via `_run_annex_command_json`).

It does not

  * strip the `normalize_paths` decorator
    I think it'd make more sense to do wholesale removal of it from methods rather than incrementally dropping it from existing methods.
  * bother using --json-error-messages for the reasons I noted above

Reply:
FTR: I disagree with your conclusion re not stripping `normalize_paths`. It silently changes the shape of the path going in an out. Personally I think there should be a single pattern: high-level code hands over path to a repo method in relative form, or absolute pointing inside `ds.repo.path`. Repo method must never try to be clever and guess how a caller might want the results and report paths in full and leave any recoding to the top-level code.

That being said, not removing it doesn't make it worse than it is, and the PR is still a massive net improvement.